[go] fix vercel dev error when a api use variables from other file (#7621)

await-ovo opened the pull request:
### Related Issues

> Related to #7621 

### 📋 Checklist

<!--
  Please keep your PR as a Draft until the checklist is complete
-->

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [x] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

await-ovo committed:
[go] fix vercel dev error (#7621)

styfle reviewed (commented):
Thanks for the PR!

Do you have an example that fails today so it can be added as a test fixture?

await-ovo commented:
yes，i have create a demo to reproduce in the related issue

await-ovo committed:
[go] add fixtures for test

await-ovo commented:
> Thanks for the PR!
> 
> Do you have an example that fails today so it can be added as a test fixture?



Thanks for the review, I've added the test fixture now(`/packages/go/test/fixtures/21-use-variable-form-other-api`), and I've changed the pattern in `util/analyze.go` to ignore the files under the `.now` and `.vercel` directories.
<img width="1379" alt="image" src="https://user-images.githubusercontent.com/41503212/160235701-227fc558-d9b1-4a1f-a7da-84141199d0f8.png">


I've also executed `yarn test-integration-once` locally, and **everything else worked except for the `fixtures/01-cowsay` fixture, but I'm confused because I'm getting the same error on the main branch**.
<img width="956" alt="Pasted Graphic 3" src="https://user-images.githubusercontent.com/41503212/160235754-bea88b72-5595-477a-8cdf-675475e0b8b1.png">


I think the code changes in this pr will only affect dev, I'm not sure where the problem is yet. I need your help, thanks!

styfle commented on packages/go/index.ts:
```diff
@@ -524,7 +538,7 @@ export async function startDevServer(
   const analyzedRaw = await getAnalyzedEntrypoint(
     workPath,
-    entrypointWithExt,
+    (files[entrypointWithExt] as FileFsRef).fsPath,
```
I think using `files` here is incorrect. Instead we should keep the call site the same and if we need the absolute path, the `getAnalyzedEntrypoint()` can do `join(workPath, entrypoint)` instead.

await-ovo replied:
@styfle  Hi, I used `files` to get the absolute path here only after reading the build function. 

The build function calls `getAnalyzedEntrypoint()` with the argument `filePath` which is the absolute path
>https://github.com/vercel/vercel/blob/main/packages/go/index.ts#L147

```
  analyzed = await getAnalyzedEntrypoint(
      workPath,
      downloadedFiles[entrypoint].fsPath,
      goModAbsPathDir
    );
```

Is it better to pass the absolute path when calling the function than to join inside the `getAnalyzedEntrypoint` function?

styfle replied:
I think it’s best to join inside the func since there are two callers so this will consolidate

styfle commented on packages/go/index.ts:
```diff
@@ -536,7 +550,16 @@ Learn more: https://vercel.com/docs/runtimes#official-runtimes/go`
+      [
+        relative(
+          goModAbsPathDir,
+          (files[entrypointWithExt] as FileFsRef).fsPath
```
Same here, instead of using `files` use `join(workPath, entrypointWithExt)`

await-ovo replied:
Ok, I tried it and it works fine, I will use `path.join` instead of `files`.

styfle commented on packages/go/index.ts:
```diff
@@ -479,13 +482,24 @@ function isReadable(v: any): v is Readable {
-  const data = await readFile(entrypoint, 'utf8');
+async function copyEntrypoint(files: string[], dest: string): Promise<void> {
+  for (const file of files) {
+    if (!existsSync(file)) {
```
Why would the file not exist?

await-ovo replied:
This may have something to do with `analyze.go`, when the directory structure is like this:
```
.
├── README.md
├── api
│   ├── date.go
│   ├── hello.go
│   ├── json.go
│   └── myinfo.go
├── go.mod
└── util
    └── util.go
```
`analyzed.watch` in  the result is like this:

```
[ 'json.go', 'api/hello.go',  'util/util.go']
```

When copying the entry and watch files, I had to ignore `json.go` which does not exist relative to `goModAbsPathDir`.

Since `analyze.go` is used by both build and dev, it works fine when building for production, so I didn't modify `analyze.go` here. Instead, I added a check for the existence of the file when copying.

styfle replied:
Hmm that seems like a bug if it returned json.go instead of api/json.go

Perhaps the working directory is wrong?

await-ovo replied:
@styfle  I changed the `analyze.go` code a bit, so that the `entrypoint` path returned in `analyzed.watch` array works.

Also, I changed `/go/test/fixtures/08-include-files/another.go`, so that the `yarn test-integration-once` does not report any errors.

await-ovo committed:
[go]: use path.join instead of getting the absolute path from files

await-ovo committed:
[go] return the entrypoint path relative to the modPath in analyzed.wach array

EndangeredMassa committed:
Merge branch 'main' into fix-go-runtime-dev-error

styfle commented on packages/go/index.ts:
```diff
@@ -479,13 +480,20 @@ function isReadable(v: any): v is Readable {
-  await writeFile(join(dest, 'entrypoint.go'), patched);
+      await outputFile(join(dest, file), patched);
+    } catch (err) {
+      debug(`copy entrypoint ${file} error`, err);
```
When would an error happen?

styfle commented on packages/go/index.ts:
```diff
@@ -479,13 +480,20 @@ function isReadable(v: any): v is Readable {
-async function copyEntrypoint(entrypoint: string, dest: string): Promise<void> {
-  const data = await readFile(entrypoint, 'utf8');
+async function copyEntrypoint(files: string[], dest: string): Promise<void> {
+  for (const file of files) {
```
Can we change this function back so the copyEntrypoint() function can be invoked multiple times? I think line 547 can be changed to a `files.map()` so you run all the copy files concurrently.

styfle commented on packages/go/test/fixtures/08-include-files/another.go:
```diff
@@ -7,7 +7,7 @@ import (
 
 // Handler function
-func Handler(w http.ResponseWriter, r *http.Request) {
+func AnotherHandler(w http.ResponseWriter, r *http.Request) {
```
Let's not change the existing tests

EndangeredMassa commented:
Closing for now, but we are tracking this internally. We'll come back to this when we next focus on Go improvements.
